File: sort_algorithm/quick_sort.go
package sort_algorithm

import (
	"fmt"
)

type QuickSort []int

func (q *QuickSort) SortImpl() {
	//*q = q.quickSort(*q, 0, len(*q)-1)
	q.QuickSort()
	fmt.Println(*q)
}

func (q *QuickSort) quickSort(arr QuickSort, l, r int) QuickSort {
	if l < r {
		mid := q.partition(arr, l, r)
		arr = q.quickSort(arr, l, mid-1)
		arr = q.quickSort(arr, mid+1, r)
	}
	return arr
}

func (*QuickSort) partition(arr []int, l, r int) int {
	pivot, i, j := arr[l], l+1, r
	for i <= j {
		for i <= j && arr[i] <= pivot {
			i++
		}
		for j >= i && arr[j] >= pivot {
			j--
		}
		if i >= j {
			break
		}
		arr[i], arr[j] = arr[j], arr[i]
	}
	arr[l], arr[j] = arr[j], pivot
	return j
}

// QuickSort 非递归版本
func (q *QuickSort) QuickSort() {
	stack := [][]int{{0, len(*q) - 1}}
	for length := len(stack); length > 0; length = len(stack) {
		l, r := stack[length-1][0], stack[length-1][1]
		stack = stack[:length-1]
		if l < r {
			mid := q.partition1(*q, l, r)
			if mid-1 > l {
				stack = append(stack, []int{l, mid - 1})
			}
			if mid+1 < r {
				stack = append(stack, []int{mid + 1, r})
			}
		}
	}
}

func (q *QuickSort) partition1(arr []int, l, r int) int {
	pivotIndex := l
	pivotValue := arr[r]

	for i := l; i < r; i++ {
		if arr[i] < pivotValue {
			arr[i], arr[pivotIndex] = arr[pivotIndex], arr[i]
			pivotIndex++
		}
	}

	arr[pivotIndex], arr[r] = arr[r], arr[pivotIndex]

	return pivotIndex
}

File: java/sort_algorithm/impl/QuickSort.java
package sort_algorithm.impl;

import sort_algorithm.Sort;

import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;

public class QuickSort implements Sort {
    @Override
    public void sort(int[] arr) {
        Deque<int[]> stack = new LinkedList<>();
        while (!stack.isEmpty()) {
            int[] x = stack.pop();
            int l = x[0], r = x[1];
            if (l < r) {
                int mid = partition(arr, l, r);
                if (mid - 1 > l) {
                    stack.push(new int[]{l, mid - 1});
                }
                if (mid + 1 < r) {
                    stack.push(new int[]{mid + 1, r});
                }
            }
        }
        System.out.println(Arrays.toString(arr));
    }
    int partition(int[] arr, int l, int r) {
        int pivot = arr[l], i = l + 1, j = r;
        while(i <= j) {
            while(i <= j && arr[i] <= pivot) i++;
            while(j >= i && arr[j] >= pivot) j--;
            if (i >= j) break;
            int tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
        }
        arr[l] = arr[j];
        arr[j] = pivot;
        return j;
    }
}
